modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,195 @@
+
+        expectedOrdinalsToConfirm <- updateAvgExpectedOrdinalsToConfirm(confirmed, globalSnapshot.ordinal)
+        _ <- updateAvgConfirmedCountPerSnapshot(confirmed)
+        _ <- updateCatchUpMode(pending, currentOrdinal, expectedOrdinalsToConfirm)
+
+        _ <- logger
+          .warn(
+            s"${pending.length} state channel binaries are still not confirmed on the global state. Waiting for the next global snapshot."
+          )
+          .whenA(pending.nonEmpty)
+      } yield ()
+
+      def processPending: F[Unit] = {
+        val retry = for {
+          pending <- pendingR.get
+          avgConfirmedPerSnapshot <- avgConfirmedCountPerSnapshotR.get
+          cap = Math
+            .round(avgConfirmedPerSnapshot.toDouble + (avgConfirmedPerSnapshot.toDouble / 2))
+            .toInt // Try sending 50% more than avg confirmed
+          toRetry = pending.take(cap)
+          _ <- toRetry.traverse(tracked => enqueue(tracked.binary))
+        } yield ()
+
+        catchUpMode.get.ifM(retry, Applicative[F].unit)
+      }
+
+      def clearPending: F[Unit] = pendingR.set(List.empty)
+
+      private def updateAvgExpectedOrdinalsToConfirm(
+        confirmedTracked: List[TrackedBinary],
+        confirmedAtOrdinal: SnapshotOrdinal
+      ): F[PosLong] = {
+        val ordinalsPassed = confirmedTracked.map(tb => confirmedAtOrdinal.value - tb.sentAtOrdinal.value)
+        val avgOrdinalsPassed = PosLong.unsafeFrom(Math.round(ordinalsPassed.sum.toDouble / ordinalsPassed.length))
+        expectedOrdinalsToConfirmR.set(avgOrdinalsPassed).as(avgOrdinalsPassed)
+      }
+
+      private def updateAvgConfirmedCountPerSnapshot(confirmedTracked: List[TrackedBinary]): F[NonNegLong] =
+        avgConfirmedCountPerSnapshotR.modify { prev =>
+          val updated = NonNegLong.from(Math.round((prev.toDouble + confirmedTracked.length) / 2)).getOrElse(prev)
+          (updated, updated)
+        }
+
+      private def updateCatchUpMode(
+        pending: List[TrackedBinary],
+        currentOrdinal: Option[SnapshotOrdinal],
+        expectedOrdinalsToConfirm: PosLong
+      ): F[Unit] =
+        catchUpMode.update(prev =>
+          if (pending.isEmpty) {
+            false
+          } else {
+            val hasStalled = pending.exists { p =>
+              val ordinalsWithoutConfirmationCount =
+                NonNegLong
+                  .from(currentOrdinal.getOrElse(SnapshotOrdinal.MinValue).value - p.sentAtOrdinal.value)
+                  .getOrElse(NonNegLong.MaxValue)
+              ordinalsWithoutConfirmationCount >= expectedOrdinalsToConfirm
+            }
+            hasStalled || prev
+          }
+        )
+
+      private def enqueue(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        for {
+          currentOrdinal <- lastGlobalSnapshotStorage.getOrdinal.map(_.getOrElse(SnapshotOrdinal.MinValue))
+          _ <- pendingR.update { pending =>
+            if (pending.exists(_.binary.hash === binary.hash)) pending
```

Comment:
Is there a valid scenario where a situation like this would happen? Seems like we shouldn't need such check.